# Babysitter and Parent Matching Database

I built a SQL database that could support a web application for babysitters to offer their services and parents to find babysitters that meet their needs. I populated fake data for the database. I did not create the web application for this project but that could be a future extension of this project. This is a course project for the University of Colorado Database Systems course (Summer 2022) within the computer science post-baccalaureate program. I demonstrate triggers and other constraints, complex queries, and indices.

## 1. Import packages and connect to sqlalchemy

In [2]:
import pandas as pd
import sqlalchemy
from sqlalchemy.types import Integer, Text, String, DateTime
from sqlalchemy import text
from sqlalchemy import create_engine
import os
import configparser

In [3]:
def GetConnection(cfg_filename) :
    mysqlcfg = configparser.ConfigParser()
    mysqlcfg.read(cfg_filename)
    dburl = mysqlcfg['mysql']['url']

    os.environ['DATABASE_URL'] = dburl  # define this env. var for sqlmagic
    print(dburl)
    eng = create_engine(dburl)
    print(eng)
    return eng.connect()

con = GetConnection("/home/jovyan/mysql.cfg")
print(con)

mysql://jest8502:6c20c190ac60b66041b9@applied-sql.cs.colorado.edu:3306/jest8502
Engine(mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502)


In [4]:
%reload_ext sql
print ("get version...")
%sql SELECT version()

get version...
1 rows affected.


[('8.0.27',)]

## 2. Load data from csv files, update the date formats, and create tables

In [5]:
# Load Data
parent = pd.read_csv('parent.csv')
child = pd.read_csv('child.csv')
babysitter = pd.read_csv('babysitter.csv')
job = pd.read_csv('job.csv')
schedule = pd.read_csv('schedule.csv')

In [6]:
# Get the dates in the job and schedule columns in a SQL Date-Friendly Format
job['date'] = pd.to_datetime(job['date'])
schedule['date'] = pd.to_datetime(schedule['date'])

In [7]:
%%sql
DROP table if exists Schedule;
DROP table if exists Job;
DROP table if exists Child;
DROP table if exists Parent;
DROP table if exists Babysitter;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.


[]

## CREATE TABLES
- There are Multiple Tables
- There are relationships between tables with Foreign Keys

The **Parent** table holds the parentID (primary key), parent first name, parent last name, zip code, and email.

The **Child** table holds the childID (primary key), parentID, and age. There is a many-to-one relationship between child and parent as a parent can have multiple children. However, each child is associated with only one parent. The parentID is the Foreign Key that references the parentID in the Parent table.

The **Job**  table would be populated by parents and hold the jobID (primary key) parentID, date of the job, job start time, job end time, and hourly pay rate. There is a many-to-one relationship between Job and Parent as a parent can post multiple jobs. However, each job corresponds to just one parent. The parentID is the Foreign Key that references the parentID in the Parent table.

The **Babysitter** table holds the babysitterID (primary key), first name, last name, zip code, minimum age of kids they can care for, max number of kids they can care for, minimum hourly pay rate, and email.

The **Schedule** table holds the babysitterID, date, startTime, endTime, and jobAccepted field. Together, the babysitterID and date fields will represent the primary key. Each row in the table will represent a day that the babysitter is available to work. The startTime and endTime are represent the hours the babysitter is available that day. I will allow for 31 calendar days for availability to show for the month of August. The jobAccepted field will hold the jobID of any “accepted jobs” on the corresponding date. There is a one-to-one relationship between the Job and Schedule tables. A particular job can be accepted by one babysitter for a particular date. There is a many-to-one relationship between the Schedule and Babysitter table as each Babysitter can list multiple dates they are available in the Schedule table. The babysitterID is a Foreign Key that references the babysitterID in the Babysitter table. The jobAccepted field is a Foreign Key that references the jobID in the Job table.

In [8]:
%%sql
DROP table if exists Parent;
CREATE TABLE Parent(
    parentID INT PRIMARY KEY AUTO_INCREMENT,
    firstName VARCHAR(100) NOT NULL,
    lastName VARCHAR(100) NOT NULL,
    email VARCHAR (100) NOT NULL,
    zipcode VARCHAR(5) NOT NULL
);

DROP table if exists Child;
CREATE TABLE Child(
    childID int AUTO_INCREMENT PRIMARY KEY,
    parentID int,
    age int,
    FOREIGN KEY (parentID) REFERENCES Parent(parentID)
);

DROP table if exists Babysitter;
CREATE TABLE Babysitter(
    babysitterID INT AUTO_INCREMENT PRIMARY KEY,
    firstName VARCHAR(100) NOT NULL,
    lastName VARCHAR(100) NOT NULL,
    email VARCHAR (100) NOT NULL,
    zipcode VARCHAR(5) NOT NULL,
    minHourlyRate DOUBLE(4, 2),
    maxNumKids INT,
    minAgeKids INT
);

DROP table if exists Job;
CREATE TABLE Job(
    jobID INT AUTO_INCREMENT PRIMARY KEY,
    parentID INT,
    date DATE,
    startTime TIME,
    endTime TIME,
    payRateHourly DOUBLE(4, 2),
    FOREIGN KEY (parentID) REFERENCES Parent(parentID)
);

DROP table if exists Schedule;
CREATE TABLE Schedule (
    babysitterID INT,
    date DATE,
    startTime TIME,
    endTime TIME,
    jobAccepted INT UNIQUE,
    FOREIGN KEY (babysitterID) REFERENCES Babysitter(babysitterID),
    FOREIGN KEY (jobAccepted) REFERENCES Job(jobID),
    PRIMARY KEY (babysitterID, date)
);


 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.


[]

## 3. CREATE TRIGGERS
**Triggers and Constraints**: I used Triggers, Foreign Keys, and NOT NULL constraints in table creation to provide rules in my data system. I created custom triggers to enforce the Babysitter’s schedule and maximum number of kids she can handle when accepting a Job (Triggers 1 & 2). I show how INSERTING new data into tables with Foreign Key attributes can trip my triggers that warn the user that they first must establish the parent, babysitter, or job profiles in the associated tables (Triggers 3-5). I also show how DELETING rows from base profile tables can trip my triggers to first DELETE associated items in other tables that use those base table attributes as Foreign Keys. A list of these triggers and other constraints are below. 

1. A babysitter cannot accept a job that conflicts with their startTime or endTime for a given date. (enforce_start_end_Time trigger)
2. A babysitter cannot accept a job to take care of more than the maximum number of kids in their profile. (enforce_maxKids) Every childID in the Child table must reference a parentID in the Parent Table. Foreign Key = Child.parentID. (Trigger 3, missing_parent_add_child” trigger, is tripped if the user tries to INSERT a child into the Child table before the corresponding parent profile is established in the Parent table)
3. Every parentID in the Job table must reference a parentID in the Parent Table. Foreign Key = Job.parentID. (Trigger 4, missing_parent_add_job” trigger, is tripped if the user tries to INSERT a job into the Job table before the corresponding parent profile is established in the Parent table
4. Every babysitterID in the Schedule table must reference a babysitterID in the Babysitter table. Foreign Key = Schedule.babysitterID. (Trigger 5, missing_babysitter_add_schedule” trigger, is tripped if the user tries to INSERT a schedule line item into the Schedule table before the corresponding babysitter profile is established in the Babysitter table
5. Every jobAccepted in the Schedule table must reference a jobID in the Job table. Foreign Key = Schedule.jobAccepted. (Trigger 6, missing_Job_add_jobAccepted” trigger, is tripped if the user tries to UDPATE the jobAccepted field in the Schedule table before the Job has been inserted into the Job table)
6. If a parent profile is DELETED from the Parent table, I first delete the associated rows for that parentID in the Child and Job tables. (Trigger 7: delete_parent)
7. If a babysitter profile is DELETED from the Babysitter table, I first delete the associated rows for that babysitterID in the Schedule table (Trigger 7: delete_babysitter)
8. The name, email, and zip code fields cannot be blank for the parents or babysitters (NOT NULL constraint when tables are created)
9. A job can only be accepted by a single Babysitter. I implemented this by making the jobAccepted field unique in the Schedule table.
10. A babysitter can only accept 1 job per day as the jobAccepted field in the Schedule table only holds a single integer value.


### Trigger #1 (enforce_start_end_Time)
If accepting a job outside of the babysitter's schedule constraints, reject

In [9]:
%%sql
drop trigger if exists enforce_start_end_Time;
create trigger enforce_scart_end_Time
BEFORE UPDATE ON Schedule
FOR EACH ROW 
BEGIN
    if exists (
        SELECT *
        FROM Schedule, Job
        WHERE Job.jobID = NEW.jobAccepted and ((Job.startTime < New.startTime) or (Job.endTime > New.endTime))
    ) THEN
        SIGNAL SQLSTATE '45000'
          SET MESSAGE_TEXT = 'Cannot accept this job. Job starts before scheduled start or end time for babysitter', MYSQL_ERRNO = 1001;
    end if;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #2 (enforce_maxKids)
If accepting a job with more kids than the babysitter can handle, reject

In [10]:
%%sql
drop trigger if exists enforce_maxKids;
create trigger enforce_maxKids
BEFORE UPDATE ON Schedule
FOR EACH ROW 
BEGIN
    if exists (
        SELECT * FROM
        
        (SELECT Parent.parentID as "ParentID", Babysitter.maxNumKids as "BabysitterMaxKids"
         FROM Babysitter, Parent, Job
         WHERE Job.jobID = NEW.jobAccepted and Job.parentID = Parent.parentID and New.BabysitterID = Babysitter.BabysitterID) jobinfo,
        
        (SELECT parentID, COUNT(parentID) as "NumKids"
        FROM Child
        GROUP BY parentID) kidinfo
        
        WHERE jobinfo.ParentID = kidinfo.parentID and jobinfo.BabysitterMaxKids < kidinfo.NumKids 
        
    ) THEN
        SIGNAL SQLSTATE '45000'
          SET MESSAGE_TEXT = 'Cannot accept this job. The family has more kids than you can handle', MYSQL_ERRNO = 1001;
    end if;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #3 (missing_parent_add_child)
Provide a useful error message when the user tries to add a child to the Child table before the corresponding parent is added to the Parent Table

In [11]:
%%sql
DROP TRIGGER IF EXISTS missing_parent_add_child;
create trigger missing_parent_add_child
BEFORE INSERT on Child
for each row
begin
  if not exists (
      SELECT parentID from Parent 
                 WHERE NEW.parentID = Parent.parentID
  ) THEN
        SIGNAL SQLSTATE '45000'
          SET MESSAGE_TEXT = 'Cannot insert a child into the database before the parent profile is established', MYSQL_ERRNO = 1001;
    end if;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #4 (missing_parent_add_job): 
Provide a useful error message when the user tries to add a job to the Job table before the corresponding babysitter is added to the Babysitter Table

In [12]:
%%sql
DROP TRIGGER IF EXISTS missing_parent_add_job;
create trigger missing_parent_add_job
BEFORE INSERT on Job
for each row
begin
  if not exists (
      SELECT parentID from Parent 
                 WHERE NEW.parentID = Parent.parentID
  ) THEN
        SIGNAL SQLSTATE '45000'
          SET MESSAGE_TEXT = 'Cannot insert a job into the database before the parent profile is established', MYSQL_ERRNO = 1001;
    end if;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #5 (missing_babysitter_add_schedule)
Provide a useful error message when the user tries to add a schedule line item to the Schedule table before the corresponding babysitter is added to the Babysitter Table

In [13]:
%%sql
DROP TRIGGER IF EXISTS missing_babysitter_add_schedule;
create trigger missing_babysitter_add_schedule
BEFORE INSERT on Schedule
for each row
begin
  if not exists (
      SELECT babysitterID from Babysitter 
                 WHERE NEW.babysitterID = Babysitter.babysitterID
  ) THEN
        SIGNAL SQLSTATE '45000'
          SET MESSAGE_TEXT = 'Cannot insert a schedule item into the database before the babysitter profile is established', MYSQL_ERRNO = 1001;
    end if;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #6 (missing_Job_add_jobAccepted)
Provide a useful error message when the user tries to update a jobAccepted item in the Schedule table when the corresponding Job is missing from the Job table

In [14]:
%%sql
DROP TRIGGER IF EXISTS missing_job_add_jobAccepted;
create trigger missing_job_add_jobAccepted
BEFORE UPDATE on Schedule
for each row
begin
  if not exists (
      SELECT jobID from Job 
                 WHERE NEW.jobAccepted = Job.jobID
  ) THEN
        SIGNAL SQLSTATE '45000'
          SET MESSAGE_TEXT = 'Cannot accept a Job that does not exist in the Jobs table', MYSQL_ERRNO = 1001;
    end if;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #7 (delete_parent)
If delete a parent, first delete the corresponding rows in the Child and Job tables

In [15]:
%%sql
DROP TRIGGER IF EXISTS delete_parent;
create trigger delete_parent
BEFORE DELETE on Parent
for each row
begin
  DELETE FROM Child where Child.parentID = OLD.parentID;
  DELETE FROM Job where Job.parentID = OLD.parentID;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

### Trigger #8 (delete_babysitter)
If delete a babysitter, first delete the corresponding rows in the Schedule table

In [16]:
%%sql
DROP TRIGGER IF EXISTS delete_babysitter;
create trigger delete_babysitter
BEFORE DELETE on Babysitter
for each row
begin
  DELETE FROM Schedule where Schedule.babysitterID = OLD.babysitterID;
end;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.
0 rows affected.


[]

## 4. CREATE INDICES
I expect that parents will frequently search for babysitters based on their pay rate, so I will make this an index. I expect that babysitters will frequently search for jobs based on their pay rate, so I will make this an index

In [17]:
%%sql 
CREATE INDEX babysitter_pay
ON Babysitter (babysitterID, minHourlyRate)

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.


[]

In [18]:
%%sql 
CREATE INDEX job_pay
ON Job (jobID, payRateHourly)

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.


[]

## 5. INSERT INITIAL DATA

In [19]:
parent.to_sql('Parent', con = con, if_exists = "append", index=False)
child.to_sql('Child', con = con, if_exists = "append", index=False)
babysitter.to_sql('Babysitter', con = con, if_exists = "append", index=False)
job.to_sql('Job', con = con, if_exists = "append", index=False)
schedule.to_sql('Schedule', con = con, if_exists = "append", index=False)

310

## 6. BASIC QUERIES TO DISPLAY CONTENTS OF TABLES

In [20]:
%sql SELECT * from Parent LIMIT 5;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
5 rows affected.


[(1, 'James', 'SMITH', 'comfortableDouglas5@blueyonder.co.uk', '90001'),
 (2, 'Robert', 'JOHNSON', 'angryLori72@club-internet.fr', '90002'),
 (3, 'John', 'WILLIAMS', 'nervousPatrick43@shaw.ca', '90003'),
 (4, 'Michael', 'BROWN', 'wide-eyedBryan@libero.it', '90004'),
 (5, 'David', 'JONES', 'Lauraspotless@yahoo.com.ar', '90005')]

In [21]:
%sql SELECT * from Child LIMIT 5;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
5 rows affected.


[(1, 1, 1), (2, 1, 2), (3, 1, 3), (4, 1, 4), (5, 2, 5)]

In [22]:
%sql SELECT * from Job LIMIT 5;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
5 rows affected.


[(1, 50, datetime.date(2022, 8, 1), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=72000), 14.0),
 (2, 51, datetime.date(2022, 8, 1), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=72000), 25.0),
 (3, 52, datetime.date(2022, 8, 1), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=72000), 7.0),
 (4, 53, datetime.date(2022, 8, 2), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=72000), 13.0),
 (5, 54, datetime.date(2022, 8, 3), datetime.timedelta(seconds=36000), datetime.timedelta(seconds=72000), 29.0)]

In [23]:
%sql SELECT * from Babysitter LIMIT 5;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
5 rows affected.


[(1, 'Louis', 'JIMENEZ', 'handsomeRaymond22@live.com', '90001', 28.0, 4, 11),
 (2, 'Mary', 'POWELL', 'comfortableCarl@yahoo.com.au', '90002', 20.0, 3, 3),
 (3, 'Patricia', 'JENKINS', 'worrisomePhilip@att.net', '90003', 9.0, 2, 4),
 (4, 'Jennifer', 'PERRY', 'Carloscareful@yahoo.fr', '90004', 24.0, 4, 4),
 (5, 'Linda', 'RUSSELL', 'nuttyKelsey21@yahoo.es', '90005', 8.0, 2, 9)]

In [24]:
%sql SELECT * from Schedule LIMIT 5;

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
5 rows affected.


[(1, datetime.date(2022, 8, 1), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=46800), None),
 (1, datetime.date(2022, 8, 2), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=50400), None),
 (1, datetime.date(2022, 8, 3), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=54000), None),
 (1, datetime.date(2022, 8, 4), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=57600), None),
 (1, datetime.date(2022, 8, 5), datetime.timedelta(seconds=36000), datetime.timedelta(seconds=61200), None)]

## 7. COMPLEX QUERIES 

### COMPLEX QUERY 1: 
Babysitters can view jobs that match their requirements. 

I want babysitters to view the job profile for jobs that match their schedule availability, minimum payrate, maximum number of kids, and location. My code walks through an example of a single babysitter (babysitterID = 3) conducting this search. She finds 2 jobs that meet her requirements and geography. The query uses the “WITH” structure for clarity and joins all the tables together: Job, Schedule, Babysitter, Parent, and Child. It also uses aggregation functions of MIN(age) and COUNT(parentIDs) and GROUP BY in the Child table to determine if the job meets the babysitter’s child count and age requirements.

Let's use babysitter #3, Patricia Jenkins as an example.  First, let's take a look at her profile. 

She lives in the 90003 zipcode, requires at least $9 per hour, she can't watch more than 2 kids, and the youngest age she can watch is age 4

In [25]:
%%sql
SELECT * 
FROM Babysitter
WHERE babysitterID=3

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(3, 'Patricia', 'JENKINS', 'worrisomePhilip@att.net', '90003', 9.0, 2, 4)]

Let's view which jobs match Patricia's requirements in her zipcode, listing the highest paying jobs first

Below, we can see there are 2 jobs that match Patricia's requirements:
-Job #24 and Job #13
-Job #24 pays the best. Let's assume that is the one she wants.

In [26]:
%%sql

WITH 
    schedulematch(babysitterID, ScheduleDate, ScheduleStart, ScheduleEnd, JobID, ParentID, JobDate, JobStart, JobEnd)
        AS (SELECT Schedule.babysitterID as "babysitterID"
            , Schedule.date as "ScheduleDate"
            , Schedule.startTime as "ScheduleStart"
            , Schedule.endTime as "ScheduleEnd"
            , Job.jobID as "JobID"
            , Job.parentID as "ParentID"
            , Job.Date as "JobDate"
            , Job.StartTime as "JobStart"
            , Job.endTime as "JobEnd"                

            FROM Job, Schedule
            WHERE Job.date = Schedule.date 
            AND Job.startTime>= Schedule.startTime
            AND Job.endTime<= Schedule.endTime
           ),
    
    paymatch(babysitterID, ParentID, JobID, JobPay, NeededPay, BabysitterZip1)
        AS (SELECT Babysitter.babysitterID as "babysitterID", Job.parentID as "ParentID", Job.jobID as "JobID"
            , Job.payRateHourly as "JobPay", Babysitter.minHourlyRate as "NeededPay", Babysitter.zipcode as "BabysitterZIP1"
            FROM Job, Babysitter
            WHERE Babysitter.minHourlyRate <= Job.payRateHourly
           ),
        
    zipmatch(babysitterID, ParentID, JobID, JobZip, ParentName, ParentEmail)
        AS (SELECT Babysitter.babysitterID as "babysitterID", Job.parentID as "ParentID", Job.jobID as "JobID", 
            Parent.zipcode as "JobZip", CONCAT(Parent.firstName, ' ', Parent.lastName) as "ParentName", Parent.email as "ParentEmail"
            FROM Job, Parent, Babysitter
            WHERE Job.parentID = Parent.parentID and Babysitter.zipcode = Parent.zipcode
           ),
        
    childmatch(ParentID, NumKids, MinAge)
        AS (SELECT parentID as "ParentID", COUNT(parentID) as "NumKids", MIN(age) as "MinAge"
            FROM Child
            GROUP BY parentID
            )        
SELECT paymatch.babysitterID, schedulematch.JobID, schedulematch.ParentID, paymatch.JobPay as "Pay Rate"
        , schedulematch.JobDate as "Job Date", schedulematch.JobStart "Start Time", schedulematch.JobEnd as "End Time"
        , zipmatch.JobZip, zipmatch.ParentName, zipmatch.ParentEmail, childmatch.NumKids as "Number of Kids in Family"
        , childmatch.MinAge as "Age of Youngest Child"
        FROM schedulematch, paymatch, zipmatch, childmatch
        WHERE schedulematch.JobID = paymatch.JobID and zipmatch.JobID = schedulematch.JobID 
        and zipmatch.babysitterID = schedulematch.babysitterID and zipmatch.JobZip = paymatch.BabysitterZip1
        and schedulematch.parentID = childmatch.ParentID
        and paymatch.babysitterID = 3
        ORDER BY JobPay DESC

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
2 rows affected.


[(3, 24, 73, 21.0, datetime.date(2022, 8, 18), datetime.timedelta(seconds=50400), datetime.timedelta(seconds=72000), '90003', 'Carl REED', 'Toniold-fashioned@sky.com', 1, 6),
 (3, 13, 3, 19.0, datetime.date(2022, 8, 7), datetime.timedelta(seconds=64800), datetime.timedelta(seconds=72000), '90003', 'John WILLIAMS', 'nervousPatrick43@shaw.ca', 1, 12)]

### COMPLEX QUERY 2: 
Parents/employers can view babysitters that might want the job

I want parents to view babysitter profiles who meet their job requirements in terms of schedule availability, minimum pay, maximum number of kids, kid age, and location. This is nearly the same query to Query #1, but displaying data tailored for the Parent. My code walks through an example of a parent searching for babysitters who might be willing to take job #24. I find 7 potential babysitters.


Let's have Carl Reed review all the babysitter options for his job \#24.
First let's look at the job details of job /#24
The job is on Aug 18 from 2-8pm and pays $21 per hour for 1 kid

In [604]:
%%sql
SELECT jobreview.jobID, jobreview.parentID, jobreview.date, jobreview.startTime, jobreview.endTime, jobreview.payRateHourly
, childreview.NumKids as "Number of Kids" from

(SELECT * 
 FROM Job
WHERE jobID=24) jobreview,

(SELECT parentID, COUNT(parentID) as "NumKids", MIN(age) as "MinAge"
FROM Child
WHERE parentID=73
GROUP BY(parentID)) childreview
 
WHERE jobreview.parentID = childreview.parentID

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(24, 73, datetime.date(2022, 8, 18), datetime.timedelta(seconds=50400), datetime.timedelta(seconds=72000), 21.0, 1)]

Now, let's see who might take the job, besides Patricia (babysitter /#13)

We can view 7 Babysitter Profiles

These are babysitters that would accept the Job's pay ($21 per hour), can handle the number of kids in the family, and can take care of the youngest child Also in the same zipcode

In [606]:
%%sql

WITH 
    schedulematch(babysitterID, ScheduleDate, ScheduleStart, ScheduleEnd, JobID, ParentID, JobDate, JobStart, JobEnd)
        AS (SELECT Schedule.babysitterID as "babysitterID"
            , Schedule.date as "ScheduleDate"
            , Schedule.startTime as "ScheduleStart"
            , Schedule.endTime as "ScheduleEnd"
            , Job.jobID as "JobID"
            , Job.parentID as "ParentID"
            , Job.Date as "JobDate"
            , Job.StartTime as "JobStart"
            , Job.endTime as "JobEnd"                

            FROM Job, Schedule
            WHERE Job.date = Schedule.date 
            AND Job.startTime>= Schedule.startTime
            AND Job.endTime<= Schedule.endTime
           ),
    
    paymatch(babysitterID, ParentID, JobID, JobPay, NeededPay, BabysitterZip1, BabysitterName, BabysitterEmail, MaxKids)
        AS (SELECT Babysitter.babysitterID as "babysitterID", Job.parentID as "ParentID"
            , Job.jobID as "JobID", Job.payRateHourly as "JobPay", Babysitter.minHourlyRate as "NeededPay", Babysitter.zipcode as "BabysitterZIP1"
            , CONCAT(Babysitter.firstName, ' ', Babysitter.lastName) as "BabysitterName", Babysitter.email as "BabysitterEmail", Babysitter.maxNumKids as "MaxKids"
            FROM Job, Babysitter
            WHERE Babysitter.minHourlyRate <= Job.payRateHourly
           ),
        
    zipmatch(babysitterID, ParentID, JobID, JobZip, ParentName, ParentEmail)
        AS (SELECT Babysitter.babysitterID as "babysitterID", Job.parentID as "ParentID", Job.jobID as "JobID", Parent.zipcode as "JobZip", CONCAT(Parent.firstName, ' ', Parent.lastName) as "ParentName", Parent.email as "ParentEmail"
            FROM Job, Parent, Babysitter
            WHERE Job.parentID = Parent.parentID and Babysitter.zipcode = Parent.zipcode
           ),
        
    childmatch(ParentID, NumKids, MinAge)
        AS (SELECT parentID as "ParentID", COUNT(parentID) as "NumKids", MIN(age) as "MinAge"
            FROM Child
            GROUP BY parentID
            )        
SELECT paymatch.babysitterID, schedulematch.JobID, paymatch.babysitterName, paymatch.babysitterEmail, paymatch.NeededPay, paymatch.MaxKids
        FROM schedulematch, paymatch, zipmatch, childmatch
        WHERE schedulematch.JobID = paymatch.JobID and zipmatch.JobID = schedulematch.JobID 
        and zipmatch.babysitterID = schedulematch.babysitterID and zipmatch.JobZip = paymatch.BabysitterZip1
        and schedulematch.parentID = childmatch.ParentID
        and paymatch.JobID = 24
        ORDER BY paymatch.babysitterID ASC

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
7 rows affected.


[(3, 24, 'Patricia JENKINS', 'worrisomePhilip@att.net', 9.0, 2),
 (13, 24, 'Nancy FISHER', 'glamorousMarco14@comcast.net', 15.0, 2),
 (33, 24, 'Amy TRAN', 'Emmanuelrich@yahoo.com.ar', 19.0, 5),
 (43, 24, 'Katherine FERNANDEZ', 'Mariehomely@libero.it', 21.0, 1),
 (53, 24, 'Ruth WEBB', 'Migueluninterested@centurytel.net', 16.0, 6),
 (73, 24, 'Ann PALMER', 'Carlylazy@yahoo.com.ar', 15.0, 5),
 (83, 24, 'Judy MILLS', 'worrisomeMarvin42@frontiernet.net', 21.0, 6)]

#### UPDATE TABLES based on Complex Query 2
The family hires Patricia Jenkins
Let's update Particia's schedule (babysitter #3 accepts job #24)

In [27]:
%%sql
UPDATE Schedule
SET jobAccepted = 24
WHERE babysitterID=3 and date = '2022-08-18'

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[]

Patricia also accepts job #13 for Aug 7th

In [28]:
%%sql
UPDATE Schedule
SET jobAccepted = 13
WHERE babysitterID=3 and date = '2022-08-07'

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[]

#### Basic Query
Let's Review Patricia's Schedule. 2 of her days are now filled up

In [29]:
%%sql
SELECT * from Schedule 
WHERE babysitterID=3 and jobAccepted IS NOT NULL

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
2 rows affected.


[(3, datetime.date(2022, 8, 7), datetime.timedelta(seconds=39600), datetime.timedelta(seconds=75600), 13),
 (3, datetime.date(2022, 8, 18), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=72000), 24)]

### COMPLEX QUERY 3
Allow a babysitter to calculate their total earnings for all the jobs they accepted in their Schedule

I want babysitters to be able to calculate their total earnings across all the jobs they accepted (as noted by the jobAccepted field in the Schedule table). I joined the Job and Schedule tables for this. I calculated total earnings by multiplying the number of hours worked (endTime - startTime) by the pay rate for all accepted jobs. 

Let's see how much Patricia will make this month if she does both these jobs.
The result is $164 for the month

In [30]:
%%sql
SELECT SUM(paycalc.hours * jobpay.payRateHourly) as "Monthly Income" FROM

(SELECT Job.startTime, Job.endTime, (Job.endTime - Job.startTime)/10000 as "hours", Schedule.jobAccepted, Schedule.babysitterID
from Job, Schedule
WHERE Job.jobID = Schedule.jobAccepted) paycalc, 

(SELECT jobID, payRateHourly
 from Job) jobpay

WHERE jobpay.jobID = paycalc.jobAccepted and paycalc.babysitterID=3


 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(164.0,)]

## 8. Trigger Tests

### Schedule Constraint
Let's test out the schedule constraint (via a trigger) with Babysitter #2, Mary Powell.
Let's look at Mary's schedule for August 1, the day she wants to take Job #1. On August 1, she is available from to 9am to 8pm at night

In [31]:
%%sql
SELECT Schedule.babysitterID, CONCAT(Babysitter.firstName, ' ', Babysitter.LastName) as "Babysitter Name"
, Schedule.date, Schedule.startTime, Schedule.endTime 
FROM Schedule, Babysitter
WHERE Schedule.babysitterID=2 and Schedule.date = '2022-08-01' and Schedule.babysitterID = Babysitter.babysitterID

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(2, 'Mary POWELL', datetime.date(2022, 8, 1), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=72000))]

Let's look at the schedule details for job #1. Oh no, the job starts at 6am. That is earlier than she can start that day

In [623]:
%%sql
SELECT * from Job
WHERE jobID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(1, 50, datetime.date(2022, 8, 1), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=72000), 14.0)]

What happens when Mary tries to add the job to her schedule? 
**It is rejected!**

In [625]:
%%sql
UPDATE Schedule
SET jobAccepted = 1
WHERE babysitterID=2 and date = '2022-08-01'

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
(MySQLdb._exceptions.OperationalError) (1001, 'Cannot accept this job. Job starts before scheduled start or end time for babysitter')
[SQL: UPDATE Schedule
SET jobAccepted = 1
WHERE babysitterID=2 and date = '2022-08-01']
(Background on this error at: https://sqlalche.me/e/14/e3q8)


### Max Kids Constraint
Let test out the max kids constraint (via a trigger) with Babysitter #9, Jessica Butler

First, let's look at Jessica's profile
We can see that she can only handle 1 kid

In [32]:
%%sql
SELECT babysitterID, CONCAT(Babysitter.firstName, ' ', Babysitter.LastName) as "Babysitter Name", maxNumKids as "Number of Kids I can Handle"
FROM Babysitter 
WHERE babysitterID=9

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(9, 'Jessica BUTLER', 1)]

Jessica wants to accept Job #11
Let's look at the details of Job #11
Oh no, Job #11 has 4 kids

In [33]:
%%sql
SELECT Job.jobID, Parent.parentID as "ParentID", CONCAT(Parent.firstName, ' ', Parent.lastName) as "Parent Name", COUNT(Child.parentID) as "Number of Kids in the Family", Job.date, Job.startTime, Job.endTime
FROM Parent, Job, Child
WHERE Job.parentID = Parent.parentID and Parent.parentID = Child.parentID and Job.jobID=11
GROUP BY Child.parentID

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(11, 1, 'James SMITH', 4, datetime.date(2022, 8, 7), datetime.timedelta(seconds=57600), datetime.timedelta(seconds=72000))]

What happens when Jessica tries to accept the job?
**She is rejected**

In [34]:
%%sql
UPDATE Schedule
SET jobAccepted = 11
WHERE babysitterID=9 and date = '2022-08-07'

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
(MySQLdb._exceptions.OperationalError) (1001, 'Cannot accept this job. The family has more kids than you can handle')
[SQL: UPDATE Schedule
SET jobAccepted = 11
WHERE babysitterID=9 and date = '2022-08-07']
(Background on this error at: https://sqlalche.me/e/14/e3q8)


### ParentID Foreign Key Constraint
Let's try to add a Child without first inserting their parent in the Parent Table
We can see that this throws the missing_parent trigger and the associated trigger message.
The parentID is a Foreign Key in the Child Table.

In [35]:
%sql INSERT INTO Child (parentID, age)  VALUES(200, 3);

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
(MySQLdb._exceptions.OperationalError) (1001, 'Cannot insert a child into the database before the parent profile is established')
[SQL: INSERT INTO Child (parentID, age)  VALUES(200, 3);]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


### ParentID Foreign Key Constraint AGAIN
Let's try to delete a parent record without first deleting the children's records
This trips the delete_parent trigger, which automatically deletes rows with the corresponding parentID in the Child and Jobs table first.
The parentID in the Parent table is a Foreign Key for the child Table
The parentID is also a Foreign Key in the Jobs Table.

First view the parent /#1 record in the Child and Job Table

In [36]:
%%sql
SELECT * from Child WHERE parentID=11

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(16, 11, 11)]

In [37]:
%%sql
SELECT * from Job WHERE parentID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(11, 1, datetime.date(2022, 8, 7), datetime.timedelta(seconds=57600), datetime.timedelta(seconds=72000), 10.0)]

Now delete parent /#1. 

In [38]:
%%sql
DELETE FROM Parent WHERE parentID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[]

Now we can see that parent #1 no longer exists in the Child or Job Table

In [39]:
%%sql
SELECT * from Child WHERE parentID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.


[]

In [40]:
%%sql
SELECT * from Job WHERE parentID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.


[]

### BabysitterID Foreign Key Constraint
Similarly, let's try to delete a Babysitter #1 without first deleting all that Babysitter's records from the Schedule table. The babysitterID in the Babysitter table is a Foreign Key in the Schedule table. This trips our trigger delete_babysitter, deleting that Babysitter's records first from the Schedule Table

First let's view Babysitter #1 in the Babysitter and Schedule Tables

In [41]:
%%sql
SELECT * from Babysitter where babysitterID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[(1, 'Louis', 'JIMENEZ', 'handsomeRaymond22@live.com', '90001', 28.0, 4, 11)]

In [42]:
%%sql
SELECT * from Schedule where babysitterID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
31 rows affected.


[(1, datetime.date(2022, 8, 1), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=46800), None),
 (1, datetime.date(2022, 8, 2), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=50400), None),
 (1, datetime.date(2022, 8, 3), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=54000), None),
 (1, datetime.date(2022, 8, 4), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=57600), None),
 (1, datetime.date(2022, 8, 5), datetime.timedelta(seconds=36000), datetime.timedelta(seconds=61200), None),
 (1, datetime.date(2022, 8, 6), datetime.timedelta(seconds=39600), datetime.timedelta(seconds=64800), None),
 (1, datetime.date(2022, 8, 7), datetime.timedelta(seconds=43200), datetime.timedelta(seconds=68400), None),
 (1, datetime.date(2022, 8, 8), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=72000), None),
 (1, datetime.date(2022, 8, 9), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=75600), None),
 (1, datetime.date(2022, 8, 10), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=79200), None),
 (1, datetime.date(2022, 8, 11), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=82800), None),
 (1, datetime.date(2022, 8, 12), datetime.timedelta(seconds=36000), datetime.timedelta(0), None),
 (1, datetime.date(2022, 8, 13), datetime.timedelta(seconds=39600), datetime.timedelta(seconds=46800), None),
 (1, datetime.date(2022, 8, 14), datetime.timedelta(seconds=43200), datetime.timedelta(seconds=50400), None),
 (1, datetime.date(2022, 8, 15), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=54000), None),
 (1, datetime.date(2022, 8, 16), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=57600), None),
 (1, datetime.date(2022, 8, 17), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=61200), None),
 (1, datetime.date(2022, 8, 18), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=64800), None),
 (1, datetime.date(2022, 8, 19), datetime.timedelta(seconds=36000), datetime.timedelta(seconds=68400), None),
 (1, datetime.date(2022, 8, 20), datetime.timedelta(seconds=39600), datetime.timedelta(seconds=72000), None),
 (1, datetime.date(2022, 8, 21), datetime.timedelta(seconds=43200), datetime.timedelta(seconds=75600), None),
 (1, datetime.date(2022, 8, 22), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=79200), None),
 (1, datetime.date(2022, 8, 23), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=82800), None),
 (1, datetime.date(2022, 8, 24), datetime.timedelta(seconds=28800), datetime.timedelta(0), None),
 (1, datetime.date(2022, 8, 25), datetime.timedelta(seconds=32400), datetime.timedelta(seconds=46800), None),
 (1, datetime.date(2022, 8, 26), datetime.timedelta(seconds=36000), datetime.timedelta(seconds=50400), None),
 (1, datetime.date(2022, 8, 27), datetime.timedelta(seconds=39600), datetime.timedelta(seconds=54000), None),
 (1, datetime.date(2022, 8, 28), datetime.timedelta(seconds=43200), datetime.timedelta(seconds=57600), None),
 (1, datetime.date(2022, 8, 29), datetime.timedelta(seconds=21600), datetime.timedelta(seconds=61200), None),
 (1, datetime.date(2022, 8, 30), datetime.timedelta(seconds=25200), datetime.timedelta(seconds=64800), None),
 (1, datetime.date(2022, 8, 31), datetime.timedelta(seconds=28800), datetime.timedelta(seconds=68400), None)]

Now Delete babysitter #1. This trips the trigger

In [43]:
%%sql
DELETE from Babysitter WHERE babysitterID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
1 rows affected.


[]

Now we can see that babysitter #1 does not exist in the Babysitter or Schedule tables

In [44]:
%%sql
SELECT * from Schedule where babysitterID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.


[]

In [45]:
%%sql
SELECT * from Babysitter where babysitterID=1

 * mysql://jest8502:***@applied-sql.cs.colorado.edu:3306/jest8502
0 rows affected.


[]